java/ql/src/Security/CWE/CWE-338/JHipsterGeneratedPRNG.ql

```diff
@@ -0,0 +1,48 @@
+/**
+ * @name Detect JHipster Generator Vulnnerability CVE-2019-16303
+ * @description Detector for the CVE-2019-16303 vulnerability that existed in the JHipster code generator.
+ * @kind problem
+ * @problem.severity error
+ * @precision very-high
+ * @id java/jhipster-prng
+ * @tags security
+ *       external/cwe/cwe-338
+ */
+
+import java
+import semmle.code.java.frameworks.apache.Lang
+
+private class PredictableApacheRandomStringUtilsMethod extends Method {
+  PredictableApacheRandomStringUtilsMethod() {
+    this.getDeclaringType() instanceof TypeApacheRandomStringUtils
+  }
+}
+
+private class PredictableApacheRandomStringUtilsMethodAccess extends MethodAccess {
+  PredictableApacheRandomStringUtilsMethodAccess() {
+    this.getMethod() instanceof PredictableApacheRandomStringUtilsMethod and
+    // The one valid use of this type that uses SecureRandom as a source of data.
+    not this.getMethod().getName() = "random"
+  }
+}
+
+private class VulnerableJHipsterRandomUtilClass extends Class {
+  VulnerableJHipsterRandomUtilClass() { getName() = "RandomUtil" }
+}
+
+private class VulnerableJHipsterRandomUtilMethod extends Method {
+  VulnerableJHipsterRandomUtilMethod() {
+    this.getDeclaringType() instanceof VulnerableJHipsterRandomUtilClass and
+    this.getName().matches("generate%") and
+    this.getReturnType() instanceof TypeString and
+    exists(ReturnStmt s, PredictableApacheRandomStringUtilsMethodAccess access |
+      s = this.getBody().(SingletonBlock).getStmt()
+    |
+      s.getResult() = access
+    )
+  }
+}
+
+from VulnerableJHipsterRandomUtilMethod the_method
+select the_method,
```

Comment:
I think the alert message should ideally explain what the actual problem is. We can't assume that all developers know what CVE numbers are. Perhaps something like `Weak random number generator used in security sensitive method`.
 
@felicitymay What do you think? Do you have suggestions for better wording?

Reply:
Happy to change it, I think your suggestion makes sense. I think mentioning the CVE somewhere makes sense.

Which is this better?

- `CVE-2019-16303: Weak random number generator used in security sensitive method`.
- `Weak random number generator used in security sensitive method. Caused by JHipster CVE-2019-16303`

Or some other option?

Reply:
I have a slight preference for the second, however, having two sentences in the alert message is perhaps not that great. @felicitymay do you have a good idea for this?

Reply:
I don't think we have any particular difficulty in using two sentences in an alert message. The other option would be to have:

```
Weak random number generator used in security sensitive method (JHipster CVE-2019-16303).

Reply:
I believe this is now resolved. I went with @felicitymay's suggestion. I'll let one of you mark this as 'resolved' if you think it is though.